File: BOJ(Java 공부용)/15649.java
import java.io.*;
import java.util.ArrayList;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int m = Integer.parseInt(input[1]);
        ArrayList<Integer> v = new ArrayList<>();
        boolean[] visited = new boolean[n];
        go(0,0,n,m,v,visited);
        bw.flush();
        bw.close();
    }
    private static void go(int here,int cnt,int n,int depth,ArrayList<Integer> v,boolean[] visited) throws IOException {
        if(cnt == depth) {
            for(int i=0;i<v.size();i++) {
                bw.write(v.get(i)+" ");
            }
            bw.write("\n");
            return;
        }
        for(int i=0;i<n;i++) {
            if(!visited[i]) {
                visited[i] = true;
                v.add(i+1);
                go(here+1,cnt+1,n,depth,v,visited);
                v.remove(v.size()-1);
                visited[i] = false;
            }
        }
    }
}

File: BOJ(Kotlin)/15649.kt
import com.sun.org.apache.xpath.internal.operations.Bool
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val gox = intArrayOf(-1,1,0,0)
val goy = intArrayOf(0,0,1,-1)
val bw = BufferedWriter(OutputStreamWriter(System.out))
fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()
    val array = mutableListOf<Int>()
    val visited = BooleanArray(n)
    dfs(n,m,0,array,visited)
    bw.flush()
    bw.close()
}

fun dfs(n : Int, m : Int,count : Int, array : MutableList<Int>, visited : BooleanArray)  {
    if (count == m) {
        for (value in array) {
            bw.write("$value ")
        }
        bw.write("\n")
        return
    }

    for (i in 0 until n) {
        if (!visited[i]) {
            visited[i] = true
            array.add(i+1)
            dfs(n,m,count+1,array,visited)
            array.removeLast()
            visited[i] = false
        }
    }

}